porch/controllers/remoterootsync/pkg/applyset/comparison.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package applyset
+
+import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog/v2"
+)
+
+// objectsAreEquals returns true if the two objects are equal.
+// If there is ambiguity / failure, we return false.
+// This method is used to determine whether it is safe to reuse the cached state when replacing objects.
+func objectsAreEqual(l, r *unstructured.Unstructured) bool {
+	if l == nil {
+		return r == nil
+	}
+	if r == nil {
+		return l == nil
+	}
+	return mapsAreEqual(l.Object, r.Object)
+}
+
+func mapsAreEqual(l, r map[string]interface{}) bool {
+	if len(l) != len(r) {
+		return false
+	}
+	for k, lv := range l {
+		rv := r[k]
+		if !interfacesAreEqual(lv, rv) {
+			return false
+		}
+	}
+	return true
+}
+
+func slicesAreEqual(l, r []interface{}) bool {
+	if len(l) != len(r) {
+		return false
+	}
+	for i, lv := range l {
+		rv := r[i]
+		if !interfacesAreEqual(lv, rv) {
+			return false
+		}
+	}
+	return true
+}
+
+func interfacesAreEqual(l, r interface{}) bool {
+	if l == nil {
+		return r == nil
+	}
+	if r == nil {
+		return l == nil
+	}
+	switch lv := l.(type) {
+	case map[string]interface{}:
+		rv, ok := r.(map[string]interface{})
+		if !ok {
+			return false
+		}
+		return mapsAreEqual(lv, rv)
+	case []interface{}:
+		rv, ok := r.([]interface{})
+		if !ok {
+			return false
+		}
+		return slicesAreEqual(lv, rv)
+	case string:
+		rv, ok := r.(string)
+		if !ok {
+			return false
+		}
+		return lv == rv
+	case int64:
+		rv, ok := r.(int64)
+		if !ok {
+			return false
+		}
+		return lv == rv
+	default:
+		// TODO: Drop to Warningf
```

Comment:
Tricky. not sure if we have guarantee that we'll be seeing only int64 types for numbers, that seems like a common case to encounter depending on yaml parser(s). See another question re: reflect.DeepEqual which may help. Otherwise I'd probably downgrade to error so user input doesn't crash the controller.